File: EndGate/EndGate.Core/Utilities/Looper.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;

namespace EndGate.Core.Utilities
{
    internal class Looper : IDisposable
    {
        private object _updateLock;
        private long _running;

        // Used to stop callback execution on callback additions or looper stops (within a Callback itself)
        private bool _callbacksModified;
        private List<LooperCallback> _callbacks;

        public Looper()
        {
            // Default values
            _updateLock = new object();
            _running = 0;
            _callbacksModified = false;
            _callbacks = new List<LooperCallback>();
        }

        public void AddCallback(LooperCallback callback)
        {
            lock (_updateLock)
            {
                _callbacks.Add(callback);
                _callbacksModified = true;
            }
        }

        public void RemoveCallback(LooperCallback callback)
        {
            lock (_updateLock)
            {
                _callbacks.Remove(callback);
                _callbacksModified = true;
            }
        }

        public void Start()
        {
            ThreadPool.QueueUserWorkItem(Run);
        }

        private void Run(object state)
        {
            var timer = Stopwatch.StartNew();
            long elapsedTime;

            if (Interlocked.Exchange(ref _running, 1) == 0)
            {
                while (true)
                {
                    // This will never have contention unless the user is setting the fps, callback or stopping the looper
                    lock (_updateLock)
                    {
                        // Verify we're still running
                        if (Interlocked.Read(ref _running) == 0)
                        {
                            break;
                        }

                        _callbacksModified = false;

                        foreach (LooperCallback callback in _callbacks)
                        {
                            elapsedTime = timer.ElapsedMilliseconds - callback.LastTriggered;

                            if (elapsedTime >= callback.TriggerFrequency)
                            {
                                callback.LastTriggered = timer.ElapsedMilliseconds;
                                callback.Callback();

                                // See if we modified 
                                if (_callbacksModified)
                                {
                                    break;
                                }
                            }
                        }
                    }

                    Thread.Yield();
                }
            }

            timer.Stop();
        }

        public void Dispose()
        {
            lock (_updateLock)
            {
                _callbacks.Clear();
                _callbacks = null;
                Interlocked.Exchange(ref _running, 0);
                _callbacksModified = true;
            }
        }
    }
}

File: EndGate/EndGate.Core.Client.JS/Utilities/Looper.js
var EndGate;
(function (EndGate) {
    (function (Core) {
        (function (Utilities) {
            var Looper = (function () {
                function Looper() {
                    this._type = "Looper";
                    this._running = false;
                    this._callbacks = [];
                }
                Looper.prototype.AddCallback = function (looperCallback) {
                    this._callbacks.push(looperCallback);
                    looperCallback.Active = true;
                    if(this._running) {
                        this.Loop(looperCallback);
                    }
                };
                Looper.prototype.RemoveCallback = function (looperCallback) {
                    var callbackFound = false, i;
                    for(i = 0; i < this._callbacks.length; i++) {
                        if(this._callbacks[i].ID === looperCallback.ID) {
                            callbackFound = true;
                            break;
                        }
                    }
                    if(callbackFound) {
                        window.clearTimeout(looperCallback.TimeoutID);
                        looperCallback.Active = false;
                        this._callbacks.splice(i, 1);
                    } else {
                        throw new Error("Callback does not exist.");
                    }
                };
                Looper.prototype.Start = function () {
                    this._running = true;
                    this.Run();
                };
                Looper.prototype.Run = function () {
                    for(var i = 0; i < this._callbacks.length; i++) {
                        this.Loop(this._callbacks[i]);
                    }
                };
                Looper.prototype.Loop = function (looperCallback) {
                    var that = this, msTimer = 1000 / looperCallback.Fps;
                    looperCallback.Callback();
                    if(looperCallback.Active) {
                        looperCallback.TimeoutID = window.setTimeout(function () {
                            that.Loop(looperCallback);
                        }, msTimer);
                    }
                };
                Looper.prototype.Dispose = function () {
                    for(var i = this._callbacks.length - 1; i >= 0; i--) {
                        this.RemoveCallback(this._callbacks[i]);
                    }
                    this._callbacks = [];
                    this._running = false;
                };
                return Looper;
            })();
            Utilities.Looper = Looper;            
        })(Core.Utilities || (Core.Utilities = {}));
        var Utilities = Core.Utilities;
    })(EndGate.Core || (EndGate.Core = {}));
    var Core = EndGate.Core;
})(EndGate || (EndGate = {}));
